service/src/event/validator.rs

```diff
@@ -0,0 +1,269 @@
+use std::collections::HashMap;
+
+use ceramic_core::StreamId;
+use ceramic_event::unvalidated::{self, init, signed};
+use ceramic_store::{CeramicOneEvent, EventInsertable, SqlitePool};
+use ceramic_validation::{cacao_verifier::Verifier, event_verifier::Verifier as _};
+use cid::Cid;
+use ipld_core::ipld::Ipld;
+use tokio::try_join;
+
+use super::{service::InvalidItem, DeliverableRequirement};
+
+use crate::Result;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum HistoryRequirement {
+    Complete,
+    InitOnly,
+    None,
+}
+
+impl From<DeliverableRequirement> for HistoryRequirement {
+    fn from(value: DeliverableRequirement) -> Self {
+        match value {
+            DeliverableRequirement::Immediate => HistoryRequirement::Complete,
+            DeliverableRequirement::Asap => HistoryRequirement::InitOnly,
+            DeliverableRequirement::Lazy => HistoryRequirement::None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EventValidator<'a> {
+    pool: &'a SqlitePool,
+    history_req: HistoryRequirement,
+}
+
+impl<'a> EventValidator<'a> {
+    pub fn new(pool: &'a SqlitePool, history_req: HistoryRequirement) -> Self {
+        Self { pool, history_req }
+    }
+
+    pub(crate) async fn validate_events(
+        &self,
+        parsed_events: Vec<EventInsertable>,
+        invalid_events: &mut Vec<InvalidItem>,
+    ) -> Result<(Vec<EventInsertable>, Vec<EventInsertable>)> {
```

Comment:
can we comment what the two different return vectors are?  And frankly just add some doc comments for this whole function

Reply:
yeah, I was planning to use a struct return here. I first wanted to poc things and makes sure the validation traits etc seem reasonable and it should actually work.. but this is definitely a hack job 😅  

I'll work on making this more reasonable today, if you have ideas on how to represent the events without needing to match and unreachable!() time/init, that'd be appreciated.. not sure if we need to add something that consumes the unvalidated event, or put arcs around some fields.. right now we get references and can't move/clone things easily.